actors/builtin/miner/miner_actor.go

```diff
@@ -481,6 +461,185 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
+				active, err := partitionSnapshot.ActiveSectors()
+				if err != nil {
+					return err
+				}
+				err = faults.Add(partIdx, active)
+				if err != nil {
+					return err
+				}
+
+				// Record faulty power for penalties.
+				//
+				// NOTE: This also includes power that was
+				// activated at the end of the last challenge
+				// window, and power from sectors that have since
+				// expired.
+				faultyPower = faultyPower.Add(partitionSnapshot.ActivePower())
+				return nil
+			})
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions")
+
+			allSectorsNos, err := bitfield.MultiMerge(allSectors...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge sector bitfields")
+
+			allIgnoredNos, err := bitfield.MultiMerge(allIgnored...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge fault bitfields")
+
+			// Load sectors as seen at the end of the last proving period.
+			sectorsSnapshot, err := LoadSectors(store, dlCurrent.SectorsSnapshot)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+			sectorInfos, err := sectorsSnapshot.LoadForProof(allSectorsNos, allIgnoredNos)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors for post challenge")
+
+			// Find the proving period start for the deadline in question.
+			ppStart := st.ProvingPeriodStart
+			if st.CurrentDeadline < params.Deadline {
+				ppStart -= WPoStProvingPeriod
+			}
+
+			targetDeadline := NewDeadlineInfo(ppStart, params.Deadline, currEpoch)
+
+			// Fails if validation succeeds.
+			challengeWindowedPost(rt, targetDeadline.Challenge, sectorInfos, post.Proofs)
+
+			// Ok, so, here's the tricky part. Now we need to mark these all as _bad_. However, compaction could have
+			// happened, so the sectors could have moved.
+			//
+			// We've addressed this in two ways:
+			//
+			// 1. Compaction cannot be done for Finality epochs after a challenge window.
+			// 2. Successful challenges will always penalize, even if we can't find the target sectors and mark them faulty.
+			//    You can run but you can't hide.
+
+			// Load sectors. We can't use the snapshot as a sector
+			// with an extended expiration could end up being
+			// scheduled to expire at the wrong time.
+			//
+			// Unfortunately, this means we'll need to load all
+			// sector infos twice (once for the proof, once to mark
+			// faulty). We can't use memoized power info because
+			// sectors may have been terminated.
+			sectors, err := LoadSectors(store, st.Sectors)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+			faultExpirationEpoch := targetDeadline.Last() + FaultMaxAge
+			powerDelta, err = dlCurrent.RecordFaults(store, sectors, info.SectorSize, QuantSpecForDeadline(targetDeadline), faultExpirationEpoch, faults)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to declare faults")
```

Comment:
If `faults` contains a partition index that no longer exists due to compaction, this will return an error and abort here.

I understand that the pause for `ChainFinality` in `CompactPartitions` exists to address this, but if this doesn't cut it, the fraudulent miner can get away without being penalized.

Reply:
The same goes for the case where the partition in question no-longer contains the faulty sectors.

> 2. Successful challenges will always penalize, even if we can't find the target sectors and mark them faulty.

Is the above comment implemented somehow?

Reply:
Ah, I see. I was thinking about sector numbers, but you're right about partition indices. I'm just going to make compaction exclusive with the challenge window.